src/Geta.NotFoundHandler/Data/SqlSuggestionRepository.cs

```diff
@@ -23,33 +23,37 @@ public SqlSuggestionRepository(IDataExecutor dataExecutor)
         public IEnumerable<SuggestionSummary> GetAllSummaries()
         {
             var summaries = new List<SuggestionSummary>();
-            var table = GetAllSuggestionCount();
+            var table = GetAllSuggestions();
 
-            foreach (DataRow row in table.Rows)
+            var suggestionsCountTable = table.DefaultView.ToTable(true, "OldUrlCount", "OldUrl");
+
+            foreach (DataRow row in suggestionsCountTable.Rows)
             {
-                var oldUrl = row[0].ToString();
+                var oldUrl = row["OldUrl"].ToString();
+                
                 var summary = new SuggestionSummary
                 {
-                    OldUrl = oldUrl, Count = Convert.ToInt32(row[1]), Referers = GetReferers(oldUrl).ToList()
+                    OldUrl = oldUrl, Count = Convert.ToInt32(row["OldUrlCount"]), Referers = GetReferers(oldUrl, table).ToList()
                 };
                 summaries.Add(summary);
             }
 
             return summaries;
         }
 
-        private IEnumerable<RefererSummary> GetReferers(string url)
+        private IEnumerable<RefererSummary> GetReferers(string url, DataTable table)
         {
             var referers = new List<RefererSummary>();
 
-            var table = GetSuggestionReferers(url);
-            if (table == null) return referers;
+            //note DataTable is not vulnerable to sql-injection since it's an in-memory object
+            var referrersTable = table.Select($"OldUrl = '{url.Replace("'", "''")}'" ,"RefererCount DESC");
+            if (referrersTable.Length == 0) return referers;
 
             var unknownReferers = 0;
-            foreach (DataRow row in table.Rows)
+            foreach (var row in referrersTable)
             {
-                var referer = row[0].ToString() ?? string.Empty;
-                var count = Convert.ToInt32(row[1].ToString());
+                var referer = row["Referer"].ToString() ?? string.Empty;
+                var count = Convert.ToInt32(row["RefererCount"].ToString());
```

Comment:
Is `ToString` call needed? I see that in another place (line 36) not.

Reply:
As you can see I just changed the row index ... so actually I have no idea and just left the rest of the logic as is ...

Reply:
ok